src/coreclr/jit/disasm.cpp

```diff
@@ -1544,6 +1562,175 @@ void DisAssembler::disOpenForLateDisAsm(const char* curMethodName, const char* c
+    coreDisToolsLibrary = coreCLRLoadedPath;
+#endif // TARGET_UNIX
+
+    HMODULE hCoreDisToolsLib = ::LoadLibraryExW(coreDisToolsLibrary, NULL, 0);
+    if (hCoreDisToolsLib == 0)
+    {
+        printf("LoadLibrary(%s) failed (0x%08x)", MAKEDLLNAME_A("coredistools"), ::GetLastError());
+        return false;
+    }
+
+    g_PtrNewDisasm = (NewDisasm_t*)::GetProcAddress(hCoreDisToolsLib, "NewDisasm");
+    if (g_PtrNewDisasm == nullptr)
+    {
+        printf("GetProcAddress 'NewDisasm' failed (0x%08x)", ::GetLastError());
+        return false;
+    }
+    g_PtrDumpInstruction = (DumpInstruction_t*)::GetProcAddress(hCoreDisToolsLib, "DumpInstruction");
+    if (g_PtrDumpInstruction == nullptr)
+    {
+        printf("GetProcAddress 'DumpInstruction' failed (0x%08x)", ::GetLastError());
+        return false;
+    }
+    g_PtrFinishDisasm = (FinishDisasm_t*)::GetProcAddress(hCoreDisToolsLib, "FinishDisasm");
+    if (g_PtrFinishDisasm == nullptr)
+    {
+        printf("GetProcAddress 'FinishDisasm' failed (0x%08x)", ::GetLastError());
+        return false;
+    }
+
+    TargetArch coreDisTargetArchitecture = Target_Host;
+
+#if defined(TARGET_ARM64)
+    coreDisTargetArchitecture = Target_Arm64;
+#elif defined(TARGET_ARM)
+    coreDisTargetArchitecture = Target_Thumb;
+#elif defined(TARGET_X86)
+    coreDisTargetArchitecture = Target_X86;
+#elif defined(TARGET_AMD64)
+    coreDisTargetArchitecture = Target_X64;
+#else
+#error Unsupported target for LATE_DISASM with USE_COREDISTOOLS
+#endif
+
+    corDisasm = (*g_PtrNewDisasm)(coreDisTargetArchitecture, &CorPrinter);
+
+    disAsmFileCorDisTools = nullptr;
+
+    return true;
+}
+
+void DisAssembler::DisasmBuffer(FILE* pfile, bool printit)
+{
+    disAsmFileCorDisTools = pfile;
+
+    unsigned              errorCount    = 0;
+    static const unsigned maxErrorCount = 50;
+
+    auto disasmRegion = [&](size_t executionAddress, size_t blockAddress, size_t blockSize) {
+        uint8_t* address       = (uint8_t*)executionAddress;
+        uint8_t* codeBytes     = (uint8_t*)blockAddress;
+        size_t   codeSizeBytes = blockSize;
+        while (codeSizeBytes > 0)
+        {
+            size_t instrLen = (*g_PtrDumpInstruction)(corDisasm, address, codeBytes, codeSizeBytes);
+            if (instrLen == 0)
+            {
+                // error
+                ++errorCount;
+#ifdef TARGET_ARM64
+                // For arm64, with fixed-size instructions, try to recover by skipping one
+                // instruction and continuing.
+                if (codeSizeBytes >= 4)
+                {
+                    CorDisToolsLogERROR("%x: %02x %02x %02x %02x", address, codeBytes[0], codeBytes[1], codeBytes[2],
+                                        codeBytes[3]);
+                    address += 4;
+                    codeBytes += 4;
+                    codeSizeBytes -= 4;
+                }
+                if (errorCount < maxErrorCount)
```

Comment:
shouldn't this check be outside of `#ifdef TARGET_ARM64`?

Reply:
no: ARM64 is the only platform currently where there is any logic to recover from a failure to disassemble. For x86/x64, it's not clear what could be done: where would you restart? For arm32, we could skip either 2 or 4 based on an easy check whether the instruction is 2-bytes or 4-bytes, but it didn't seem worth it.